### PR TITLE
Initial docs for extension developers

### PR DESCRIPTION
**Pull request by hairyhum**

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -1,31 +1,379 @@
...
+ - `Helpers` - various functions to use in commands and main module
+ - `Config` - enviroment variable/argument values merge
+ - `Output` - output formatting logic
+ - `Parser` - OptionParser configuration
+
+### Command execution process:
+
+1. Arguments parsing
+
+ Command line arguments are parsed with OptionParser [http://elixir-lang.org/docs/stable/elixir/OptionParser.html]
```
Comment (acogoluegnes) [inline DESIGN.md]: Command line arguments are parsed with OptionParser [http://elixir-lang.org/docs/stable/elixir/OptionParser.html] => Command line arguments are parsed with [OptionParser](http://elixir-lang.org/docs/stable/elixir/OptionParser.html).

File: DESIGN.md [hunk 2]
```diff
@@ -1,31 +1,379 @@
...
+ Parser return a list of unnamed arguments and a map of options (named arguemtns)
```
Comment (acogoluegnes) [inline DESIGN.md]: Parser return a list of unnamed arguments and a map of options (named arguemtns) => Parser returns a list of unnamed arguments and a map of options (named arguments).

File: DESIGN.md [hunk 3]
```diff
@@ -1,31 +1,379 @@
...
+ First unnamed argument is a command name.
+ Named arguments can be global or command specific.
+ Command specific argument names and types are apecified in `switches/0` callback.
```
Comment (acogoluegnes) [inline DESIGN.md]: in `switches/0` callback => in the `switches/0` callback

File: DESIGN.md [hunk 4]
```diff
@@ -1,31 +1,379 @@
...
+
+ If arguments list is not empty, its first element is considered a command name.
+ Command name is converted to CamelCase and a module with
+ `RabbitMQ.CLI.*.Commands.<CommandName>Command` name is selected as a command module.
+
+ List of available command modules depend on current tool scope (see [Command scopes](#command-scopes))
+
+3. Defaults and validation
+
+ After command module is found, effective command arguments are calculated by
```
Comment (acogoluegnes) [inline DESIGN.md]: After command module is found => After the command module is found

File: DESIGN.md [hunk 5]
```diff
@@ -1,31 +1,379 @@
...
+ merging global defaults and command specific defaults for both unnamed and named arguments.
+
+ Command specify defaults using `merge_defaults/2` callback (see [Command behaviour](#command-behaviour))
```
Comment (acogoluegnes) [inline DESIGN.md]: Command specify defaults => A command specifies defaults

File: DESIGN.md [hunk 6]
```diff
@@ -1,31 +1,379 @@
...
+
+ Arguments then validated using `validate/2` callback
```
Comment (acogoluegnes) [inline DESIGN.md]: Arguments then validated using `validate/2` callback => Arguments are then validated using the `validate/2` callback.

File: DESIGN.md [hunk 7]
```diff
@@ -1,31 +1,379 @@
...
+
+4. Command execution
+
+ Command is executed with `run/2` callback, which contains main command logic.
```
Comment (acogoluegnes) [inline DESIGN.md]: Command is executed with `run/2` callback, which contains main command logic. => A command is executed in the `run/2` callback, which contains the main logic.

File: DESIGN.md [hunk 8]
```diff
@@ -1,31 +1,379 @@
...
+ This callback can return any value.
+
+ Output from `run/2` callback is being processed with `output/2` callback,
```
Comment (acogoluegnes) [inline DESIGN.md]: Output from the `run/2` callback is being processed with the `output/2` callback,

File: DESIGN.md [hunk 9]
```diff
@@ -1,31 +1,379 @@
...
+ which should format the output to specific type. (see [Output formatting](#output-formatting))
```
Comment (acogoluegnes) [inline DESIGN.md]: to specific type => to a specific type

File: DESIGN.md [hunk 10]
```diff
@@ -1,31 +1,379 @@
...
+
+ Output callback can specify return code to be returned by command
```
Comment (acogoluegnes) [inline DESIGN.md]: This sentence needs rephrasing, it's not clear. What should return the `output/2` callback?

File: DESIGN.md [hunk 11]
```diff
@@ -1,31 +1,379 @@
...
+
+5. Printing and formatting
+
+ `output/2` callback return value is being processed in `output.ex` module by
+ formatter and printer.
```
Comment (acogoluegnes) [inline DESIGN.md]: `output/2` callback return value is being processed in `output.ex` module by formatter and printer => The `output/2` callback return value is being processed in the `output.ex` module by the appropriate formatter and printer

File: DESIGN.md [hunk 12]
```diff
@@ -1,31 +1,379 @@
...
+
+ Formatter translates the output value to a sequence of strings or error value.
```
Comment (acogoluegnes) [inline DESIGN.md]: Formatter => A formatter

File: DESIGN.md [hunk 13]
```diff
@@ -1,31 +1,379 @@
...
+ Example formatters are `json`, `csv` and `erlang`
+
+ Printer renders formatted strings to output device (e.g. stdout, file)
```
Comment (acogoluegnes) [inline DESIGN.md]: A printer renders formatted strings to an output device (e.g. stdout, file)

File: DESIGN.md [hunk 14]
```diff
@@ -1,31 +1,379 @@
...
+
+6. Return
+
+ Errors during execution (e.g. validation failures, command errors) are being printed
+ to `stderr`
```
Comment (acogoluegnes) [inline DESIGN.md]: Needs dot at the end.

File: DESIGN.md [hunk 15]
```diff
@@ -1,31 +1,379 @@
...
+
+ If command has failed to execute, non-zero code is returned.
```
Comment (acogoluegnes) [inline DESIGN.md]: If command has failed to execute, non-zero code is returned => If a command has failed to execute, a non-zero code is returned

File: DESIGN.md [hunk 16]
```diff
@@ -1,31 +1,379 @@
...
+
+
+## Usage:
+
+CLI tool is an Elixir Mix application. It is compiled into escript executable file.
```
Comment (acogoluegnes) [inline DESIGN.md]: into escript => into an escript

File: DESIGN.md [hunk 17]
```diff
@@ -1,31 +1,379 @@
...
+This file embeds Elixir, rabbit_common, and rabbitmqcli applications and can be executed anywhere
+where `escript` is installed (it comes as a part of `erlang`).
+
+Although, some commands require rabbitmq broker code and data directory to be known.
+For example, commands in `rabbitmq-plugins` tool and those controlling clustering
+(e.g. `forget_cluster_node`, `rename_cluster_node` ...)
+
+Those directories can be defined using environment options or rabbitmq environment variables.
+
+In the broker distribution the escript file is called from shell/cmd sctipt, which loads
```
Comment (acogoluegnes) [inline DESIGN.md]: from shell/cmd sctipt => from an shell/cmd script

File: DESIGN.md [hunk 18]
```diff
@@ -1,31 +1,379 @@
...
+broker environment and exports it into the script.
+
+Environment variables also specify enabled plugins file and plugins directory.
```
Comment (acogoluegnes) [inline DESIGN.md]: Environment variables also specify enabled plugins file and plugins directory => Environment variables also specify enabled the locations of the plugins file and of the plugins directory

File: DESIGN.md [hunk 19]
```diff
@@ -1,31 +1,379 @@
...
+### Global arguments
+
+#### Commonly used arguments:
+
+- node (n): atom, broker node name, defaults to `rabbit@<current host>`
+- quiet (q): boolean, if set to `true`, command banner is not shown, defaults to `false`
+- timeout (t): integer, timeout value in **seconds** (used in some commands), defaults to `infinity`
+- vhost (p): string, vhost to talk to, defaults to `/`
+- formatter: string, formatter to use to format command output. (see [Output formatting](#output-formatting))
+- printer: string, printer to rende output (see [Output formatting](#output-formatting))
```
Comment (acogoluegnes) [inline DESIGN.md]: rende => render

File: DESIGN.md [hunk 20]
```diff
@@ -1,31 +1,379 @@
...
+Will result with unnamed arguments list `["list_queues", "name", "pid"]`
+and named options map `%{vhost: "my_vhost", timeout: 10, quiet: true}`
+
+
+### Usage help
+
+Usage is shown when the CLI is called without any arguments, or if there are some
+problems parsing arguments.
+
+In that cases exit code is `64`
```
Comment (acogoluegnes) [inline DESIGN.md]: Needs dot at the end.

File: DESIGN.md [hunk 21]
```diff
@@ -1,31 +1,379 @@
...
+
+If you want to show usage and return `0` exit code run `help` command
```
Comment (acogoluegnes) [inline DESIGN.md]: Needs dot at the end.

File: DESIGN.md [hunk 22]
```diff
@@ -1,31 +1,379 @@
...
+
+`rabbitmqctl help`
+
+Each tool (`rabbitmqctl`, `rabbitmq-plugins`) shows it's scope of commands (see [Command scopes](#command-scopes))
```
Comment (acogoluegnes) [inline DESIGN.md]: it's => its

File: DESIGN.md [hunk 23]
```diff
@@ -1,31 +1,379 @@
...
+If function returns `{:validation_failure, err}`
+CLI will print usage to `stderr` and exit with non-zero exit code (typically 64).
+
+`run(arguments :: List.t, options :: Map.t) :: run_result :: any`
+
+Run command. This function usually calls RPC on broker.
+
+`output(run_result :: any, options :: Map.t) :: :ok | {:ok, output :: any} | {:stream, Enum.t} | {:error, ExitCodes.exit_code, [String.t]}`
+
+Cast the return value of `run/2` command to formattable value and exit code.
```
Comment (acogoluegnes) [inline DESIGN.md]: Cast the return value of `run/2` command to formattable value and exit code => Cast the return value of the `run/2` callback to a formattable value and exit code

File: DESIGN.md [hunk 24]
```diff
@@ -1,31 +1,379 @@
...
+`{:ok, output}` - return `exit` code and print output with `format_output/2` callback in formatter
+
+`{:stream, stream}` - format with `format_stream/2` callback in formatter, iterating over enumerable elements.
+Can return non-zero code, if error occurs during stream processing.
+
+`{:error, exit_code, strings}` - print error messages to `stderr` and return `exit_code` code
+
+There is a default implementation for this callback in `DefaultOutput` module
+
+Most of commands use default implementation via `use RabbitMQ.CLI.DefaultOutput`
```
Comment (acogoluegnes) [inline DESIGN.md]: Most of commands use default implementation via => Most of the commands use the default implementation via

File: DESIGN.md [hunk 25]
```diff
@@ -1,31 +1,379 @@
...
+To add a command to `rabbitmqctl`, one should either name it with
+the `RabbitMQ.CLI.Ctl.Commands` prefix or add the `scopes()` callback,
+returning a list with `:ctl` element
+
+## Output formatting
+
+The CLI supports extensible output formatting. Formatting consists of two stages:
+
+- formatting - translating command output to a sequence of lines
+- printing - outputing the lines to some device (e.g. stdout or filesystem)
```
Comment (acogoluegnes) [inline DESIGN.md]: outputing => outputting

File: DESIGN.md [hunk 26]
```diff
@@ -1,31 +1,379 @@
...
+Formatter is a module, implementing the `RabbitMQ.CLI.FormatterBehaviour` behaviour:
+
+`format_output(output :: any, options :: Map.t) :: String.t | [String.t]`
+
+Format a single value, returned. It accepts output from command and named arguments (options)
+and returns a list of strings, that should be printed.
+
+`format_stream(output_stream :: Enumerable.t, options :: Map.t) :: Enumerable.t`
+
+Format a stream of return values. This function utilizes elixir
```
Comment (acogoluegnes) [inline DESIGN.md]: utilizes => uses

File: DESIGN.md [hunk 27]
```diff
@@ -1,31 +1,379 @@
...
+Stream [http://elixir-lang.org/docs/stable/elixir/Stream.html] abstraction
+to define processing of continious data, so the CLI can output data in realtime.
```
Comment (acogoluegnes) [inline DESIGN.md]: continious => continuous

File: DESIGN.md [hunk 28]
```diff
@@ -1,31 +1,379 @@
...
+
+Used in `list_*` commands, that emit data in asynchronously.
```
Comment (acogoluegnes) [inline DESIGN.md]: Delete "in".

File: DESIGN.md [hunk 29]
```diff
@@ -1,31 +1,379 @@
...
+A printer module performs printing. Printer module should implement
+the `RabbitMQ.CLI.PrinterBehaviour` behaviour:
+
+`init(options :: Map.t) :: {:ok, printer_state :: any} | {:error, error :: any}`
+
+Init the internal printer state (e.g. open a file handler).
+
+`finish(printer_state :: any) :: :ok`
+
+Finalize the internal printer state
```
Comment (acogoluegnes) [inline DESIGN.md]: Needs dot at the end.

File: DESIGN.md [hunk 30]
```diff
@@ -1,31 +1,379 @@
...
+
+Some commands require information about the server environment to run correctly.
+Such information is:
+
+- rabbitmq code directory
+- rabbitmq mnesia directory
+- enabled plugins file
+- plugins directory
+
+Enabled plugins and plugins directory also used to locate plugins commands.
```
Comment (acogoluegnes) [inline DESIGN.md]: Enabled plugins and plugins directory also used to locate plugins commands => Enabled plugins file and plugins directory are also used to locate plugins commands

**Comment by hairyhum** [comment]
Thanks for comments, @acogoluegnes

**Pull request merged**